contracts/token/ERC2258.sol

```diff
@@ -0,0 +1,61 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity 0.6.11;
+
+import "lib/openzeppelin-contracts/contracts/math/SafeMath.sol";
+import "lib/openzeppelin-contracts/contracts/token/ERC20/ERC20.sol";
+
+/// @title ERC2258 implements base level ERC2258 functionality for custodial functionality
+abstract contract ERC2258 is ERC20 {
+
+    using SafeMath for uint256;
+
+    mapping(address => mapping(address => uint256)) public custodyAllowance;           // Amount of funds that are "locked" at a certain address
+    mapping(address => uint256)                     public totalCustodyAllowance;      // Total amount of funds that are "locked" for a given user, cannot be greater than balance
+
+    event         CustodyTransfer(address indexed custodian, address indexed from, address indexed to, uint256 amount);
+    event CustodyAllowanceChanged(address indexed tokenHolder, address indexed custodian, uint256 oldAllowance, uint256 newAllowance);
+
+    /**
+        @dev   Increase the custody allowance for a given `custodian` corresponding to `msg.sender`.
+        @dev   It emits a `CustodyAllowanceChanged` event.
+        @param custodian Address which will act as custodian of a given `amount` for a tokenHolder.
+        @param amount    Number of FDTs custodied by the custodian.
+    */
+    function increaseCustodyAllowance(address custodian, uint256 amount) external virtual {
+        uint256 oldAllowance      = custodyAllowance[msg.sender][custodian];
+        uint256 newAllowance      = oldAllowance.add(amount);
+        uint256 newTotalAllowance = totalCustodyAllowance[msg.sender].add(amount);
+
+        require(custodian != address(0),                     "ERC2258:BAD_CUST");
+        require(amount    != uint256(0),                     "ERC2258:BAD_AMT");
+        require(newTotalAllowance <= balanceOf(msg.sender),  "ERC2258:INSUF_BAL");
```

Comment:
Note tha now, the Loan, Pool, and StakeLocker will throw ERC2258 error messages, instead of the L, P, or SL.